src/bun.js/node/types.zig

```diff
@@ -1835,548 +1849,2876 @@ pub const Emitter = struct {
+            var normalizedCwd = buf[0 ..len];
 
-        if ((ext.len != out.len or out.len == base_slice.len) and strings.endsWith(out, ext)) {
-            out = out[0 .. out.len - ext.len];
+            // Translated from the following JS code:
+            //   return StringPrototypeSlice(cwd, StringPrototypeIndexOf(cwd, '/'));
+            const index = std.mem.indexOfScalar(u8, normalizedCwd, CHAR_FORWARD_SLASH);
+            // Account for the -1 case of String#slice in JS land
+            return if (index != null) normalizedCwd[index.? ..len] else normalizedCwd[len - 1 ..len];
         }
 
-        return JSC.ZigString.init(out).withEncoding().toValueGC(globalThis);
+        // We're already on POSIX, no need for any transformations
+        return cwd;
     }
 
-    fn dirnameWindows(path: []const u8) []const u8 {
-        if (path.len == 0)
-            return ".";
-
-        const root_slice = std.fs.path.diskDesignatorWindows(path);
-        if (path.len == root_slice.len)
-            return root_slice;
-
-        const have_root_slash = path.len > root_slice.len and (path[root_slice.len] == '/' or path[root_slice.len] == '\\');
+    /// Taken from Zig 0.11.0 zig/src/resinator/rc.zig
+    /// https://github.com/ziglang/zig/blob/776cd673f206099012d789fd5d05d49dd72b9faa/src/resinator/rc.zig#L266
+    ///
+    /// Lowers ASCII values, non-ASCII values are returned directly
+    inline fn toLowerImpl(comptime T: type, a_c: T) T {
+        if (@TypeOf(T) != u16) {
+            return std.ascii.toLower(a_c);
+        }
+        return if (a_c < 128) @intCast(std.ascii.toLower(@intCast(a_c))) else a_c;
+    }
 
-        var end_index: usize = path.len - 1;
+    /// Based on Node v21.6.1 path.posix.basename:
+    /// https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L1309
+    pub fn basenamePosixImpl(comptime T: type, path: []const T, suffix: ?[]const T) []const T {
+        // validateString of `path` is performed in fn pub basename.
+        const len = path.len;
+        // Exit early for easier number type use.
+        if (len == 0) {
+            return "";
+        }
+        var start: usize = 0;
+        // We use an optional value instead of -1, as in Node code, for easier number type use.
+        var end: ?usize = null;
+        var matchedSlash: bool = true;
+
+        if (suffix != null and suffix.?.len > 0 and suffix.?.len <= len) {
+            const suffix_not_null = suffix.?;
+            if (std.mem.eql(T, suffix_not_null, path)) {
+                return "";
+            }
+            // We use an optional value instead of -1, as in Node code, for easier number type use.
+            var extIdx: ?usize = suffix_not_null.len - 1;
+            // We use an optional value instead of -1, as in Node code, for easier number type use.
+            var firstNonSlashEnd: ?usize = null;
+            var i: usize = len - 1;
+            while (i >= 0) : (i -= if (i > 0) 1 else break) {
```

Comment:
Covering the logic. You can side by side it with the Node implementation: https://github.com/nodejs/node/blob/main/lib/path.js#L1309

<img width="1556" alt="" src="https://github.com/oven-sh/bun/assets/4303/f5c15c4e-973c-4c86-8397-ef0d7d15d478">

Changing a loop structure, a if block cascade or iteration count is an easy way to introduce breaking differences.

Reply:
We could move the condition to the bottom of the loop to make it less cluttered. Here, we need to also iterate on the `i` of zero. We break because it can't go into the `-1` zone being a usize it will segfault with an int error.

Reply:
Updated with i64 to simplify loop condition.